### PR TITLE
Fix batching of events

### PR DESCRIPTION
**Pull request by bahlo**

### CODE REVIEW

File: src/worker.js [hunk 1]
```diff
@@ -18,31 +18,31 @@ const generateId = length => {
...
+      await fn.apply(context, args)
     } else {
-      lastFn = setTimeout(() => {
-        if (Date.now() - lastTime >= wait) {
-          fn.apply(context, args)
-          lastTime = Date.now()
-        }
-      }, Math.max(wait - (Date.now() - lastTime), 0))
+      lastCall = Date.now()
+      await new Promise(resolve => {
```
Comment (mhr3) [inline src/worker.js:34]: Isn't this making every single request take up to `wait` milliseconds?

Comment (bahlo) [reply src/worker.js:34]: Good catch, pushed e6d125b which clears the old timeout before setting a new one as well as making sure the timeout is properly calculated.

File: src/worker.js [hunk 2]
```diff
@@ -49,7 +58,7 @@ async function sendLogs () {
   batch = []
 
   const url = `${axiomEndpoint}/v1/datasets/${axiomDataset}/ingest`
-  return fetch(url, {
+  return withTimeout(30_000, fetch(url, {
```
Comment (mhr3) [inline src/worker.js:61]: Does CF not support `signal`?
`fetch(url, { signal: AbortSignal.timeout(30000) })`

File: src/worker.js [hunk 3]
```diff
@@ -18,35 +18,38 @@ const generateId = length => {
...
-          fn.apply(context, args)
-          lastTime = Date.now()
-        }
-      }, Math.max(wait - (Date.now() - lastTime), 0))
+      timeoutInProgress = true
+      await new Promise(resolve => {
+        setTimeout(() => {
+          fn.apply(context, args).then(() => {
+            resolve()
+            timeoutInProgress = false
```
[the remark this answers is not in the document]
Comment (bahlo) [reply src/worker.js]: Fixed

File: src/worker.js [hunk 4]
```diff
@@ -18,31 +18,25 @@ const generateId = length => {
...
-      lastFn = setTimeout(() => {
-        if (Date.now() - lastTime >= wait) {
-          fn.apply(context, args)
-          lastTime = Date.now()
-        }
-      }, Math.max(wait - (Date.now() - lastTime), 0))
+      timeoutInProgress = true
+      await new Promise(resolve => {
+        setTimeout(() => {
+          fn.apply(context, args).then(() => {
```
Comment (mhr3) [inline src/worker.js:33]: Also, .catch() to reject the outer Promise? Or perhaps just resolve/ignore so not break the original request?

Comment (bahlo) [reply src/worker.js]: Done

**Review by mhr3 — APPROVED** [review]

**Pull request merged**